DOCS/man/mpv: improve path docs and clarify config-dir path behavior (#15218)
mpv-player/mpv: DOCS/man/mpv.rst

The special argument ``--`` can be used to stop mpv from interpreting the
following arguments as options.

> For paths passed to mpv CLI options, the situation is further complicated by the

Reviewer: this is not correct, the old docs intentionally talked about "suboptions", because only then you have a problem with `,` or `:`.
if you have a simple option that takes a single string (path), you can pass anything to it without problems.

Author: Ahh, I will rephrase to match the original explanation.

wrapped in the "fixed-length" syntax, e.g. ``%n%string_of_length_n`` (see above).

When using the libmpv API, you should strictly avoid using ``mpv_command_string``
for invoking the ``loadfile`` command, and instead prefer e.g. ``mpv_command``
to avoid the need for filename escaping.

The same applies when you're using the scripting API, where you should avoid using
``mp.command``, and instead prefer using the "separate parameters" variants (such
as ``mp.commandv`` or ``mp.command_native``), so that the filenames will always be
correctly interpreted.

Some mpv options will interpret special meanings for paths starting with ``~``,
making it easy to dynamically find special directories, such as referring to the
current user's home directory or the mpv configuration directory.

When using the special ``~`` prefix, there must always be a trailing ``/`` after
the special path prefix. In other words, ``~`` doesn't work, but ``~/`` will work.

The following special paths/keywords are currently recognized:

.. warning::

    Beware that if ``--no-config`` is used, all of the "config directory"-based
    paths (``~~/``, ``~~home/`` and ``~~global/``) will resolve to empty strings.

>     This means that commands such as ``expand-path "~~home/"`` will return an

Reviewer: this is a concrete example so "such as" doesn't make sense here.
I'd rather say
> This means that `~~home/` will expand to any empty string, while `~~home/foo/bar` will expand to `foo/bar`
> (a relative path), which can be unexpected.

Author: I wanted to include an example of a command that's common and affected by this, but I'll rephrase it! :)

    empty string, and ``expand-path "~~home/foo/bar"`` will only return ``foo/bar``
    (without any path prefix), which might not be what you expected.

>     Furthermore, certain commands such as ``mp.find_config_file("mpv.conf")``

Reviewer 2: This is internal API, why is this mentioned here? Why external scripts uses it?

Author: I saw it in `lua.c` and in various internal `.lua` scripts and am using it myself to find files in the config-dirs. It does the right thing: Searches all config dirs for the given path.

I did not find any alternatives that can search all config dirs from user scripts.

Author: Anyway, I'll remove the mention of the specific command and just say that any commands that search in config dirs will fail.

    won't have any config paths to search in, and would always return ``nil``,
    which means "not found".

    Be sure that your scripts can handle the "no config" scenarios.

================ ===============================================================
Name             Meaning
================ ===============================================================
``~~/``          If the subpath exists in any of mpv's config directories, then
                 the path of the existing file/dir is returned. Otherwise this
                 is equivalent to ``~~home/``.

> ``~/``           user home directory root (similar to a shell's ``$HOME``)

Reviewer: this changes the meaning subtly.
old: similar to how it works in a shell, equivalent to `$HOME`
new: similar to how `$HOME` works in a shell
the old one is more precise

Author: The old one was really imprecise. But I'll find a new phrasing.

> ``~/``           user home directory root (equivalent to ``$HOME`` in a shell)

Author: Yeah good point, we should mention `~/` shell paths as they are the most obvious equivalent. I'll do that now.

The special argument ``--`` can be used to stop mpv from interpreting the
following arguments as options.

> For paths passed to mpv CLI suboptions (options that have multiple `:` and

Reviewer: nit: this is not specific to CLI
```suggestion
For paths passed to mpv suboptions (options that have multiple `:` and
```

Author: Good point, thanks!

`,`-separated values), the situation is further complicated by the need to
escape special characters. To work around this, the path can instead be wrapped
in the "fixed-length" syntax, e.g. ``%n%string_of_length_n`` (see above).

When using the libmpv API, you should strictly avoid using ``mpv_command_string``
for invoking the ``loadfile`` command, and instead prefer e.g. ``mpv_command``
to avoid the need for filename escaping.

The same applies when you're using the scripting API, where you should avoid using

> ``mp.command``, and instead prefer using the "separate parameters" variants (such
> as ``mp.commandv`` or ``mp.command_native``), so that the filenames will always be
> correctly interpreted.

Reviewer: i think it's clear enough without being so verbose
```suggestion
``mp.command``, and instead prefer using ``mp.commandv`` or ``mp.command_native``.
```

Some mpv options will interpret special meanings for paths starting with ``~``,
making it easy to dynamically find special directories, such as referring to the
current user's home directory or the mpv configuration directory.

When using the special ``~`` prefix, there must always be a trailing ``/`` after
the special path prefix. In other words, ``~`` doesn't work, but ``~/`` will work.

The following special paths/keywords are currently recognized:

.. warning::

    Beware that if ``--no-config`` is used, all of the "config directory"-based
    paths (``~~/``, ``~~home/`` and ``~~global/``) will be empty strings.

    This means that ``~~home/`` would expand to an empty string, and that

>     sub-paths such as ``~~home/foo/bar"`` would expand to "relative paths"
>     without any path prefix (``foo/bar``), which may not be what you expected.

Reviewer: ```suggestion
    sub-paths such as ``~~home/foo/bar"`` would expand to a 
    relative path (``foo/bar``), which may not be what you expected.
```

Author: Good change!

    Furthermore, any commands that search in config directories will fail

>     to find anything, since there won't be any directories to search in.

Reviewer: honestly not sure if we should state this, because it's so obvious

Author: I was doubtful too, but I figured it's the "Paths" section and that it's worth mentioning what happens to paths when `--no-config` is used. I don't know either. I'll leave it for now but if you have a strong opinion we can remove it next.

    Be sure that your scripts can handle these "no config" scenarios.

================ ===============================================================
Name             Meaning
================ ===============================================================

> ``~/``           The current user's home directory (equivalent to ``~/`` and
>                  ``$HOME/`` in terminal environments).

Reviewer: these two are the same thing
```suggestion
``~/``           The current user's home directory (just like ``$HOME`` in a shell).
```

Author: Yeah, but I was asked to add both examples in a previous review, with the motivation that `~/` in mpv is easier to understand when comparing it to `~/` in a shell rather than `$HOME/`. I was also asked to add the trailing slashes for extra clarity. Hehe. :D

``~/`` expands to the user's home directory. (The trailing ``/`` is always
required.) The following paths are currently recognized:

> **Important:** Beware that if ``--no-config`` is used, then all of the "config

Author: Oh thanks for pointing that out. I'll try to figure out the admonition syntax now!
